src/main/kotlin/org/rust/lang/core/macros/MacroDiagram.kt

```diff
@@ -0,0 +1,115 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.lang.core.macros
+
+import com.intellij.psi.PsiElement
+import com.intellij.psi.impl.source.tree.LeafPsiElement
+import org.rust.lang.core.psi.RsMacro
+import org.rust.lang.core.psi.RsMacroBinding
+import org.rust.lang.core.psi.RsMacroBindingGroup
+import org.rust.lang.core.psi.ext.fragmentSpecifier
+
+enum class FragmentKind {
+    Ident,
+    Path,
+    Expr,
+    Ty,
+    Pat,
+    Stmt,
+    Block,
+    Item,
+    Meta,
+    Tt,
+    Vis,
+    Literal,
+    Lifetime;
+
+    companion object {
+        fun fromString(s: String): FragmentKind = when (s) {
+            "ident" -> Ident
+            "path" -> Path
+            "expr" -> Expr
+            "ty" -> Ty
+            "pat" -> Pat
+            "stmt" -> Stmt
+            "block" -> Block
+            "item" -> Item
+            "meta" -> Meta
+            "vis" -> Vis
+            "tt" -> Tt
+            "lifetime" -> Lifetime
+            "literal" -> Literal
+            else -> error("Invalid fragment kind")
```

Comment:
`s` comes here from user code, so it can be any string in general. Choose either return null or add specific `unknown` FragmentKind.

Reply:
@vlad20012 or maybe use `Tt` for unknown kinds as the most general kind?

Reply:
I think a pattern with unknown fragment specifier should never be matched successfully, so it looks like a bad idea. I'd prefer returning null.

Reply:
OK, done